service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/CreateAzureIpStep.java

```diff
@@ -0,0 +1,95 @@
+import bio.terra.workspace.service.resource.controlled.ControlledAzureIpResource;
+import bio.terra.workspace.service.workspace.WorkspaceService;
+import bio.terra.workspace.service.workspace.model.AzureCloudContext;
+import com.azure.resourcemanager.network.models.PublicIpAddress;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CreateAzureIpStep implements Step {
+
+  private static final Logger logger = LoggerFactory.getLogger(CreateGcsBucketStep.class);
+  private final AzureConfiguration azureConfig;
+  private final CrlService crlService;
+  private final ControlledAzureIpResource resource;
+  private final WorkspaceService workspaceService;
+  private final AzureCloudContext azureCloudContext;
+
+  public CreateAzureIpStep(
+      AzureConfiguration azureConfig,
+      AzureCloudContext azureCloudContext,
+      CrlService crlService,
+      ControlledAzureIpResource resource,
+      WorkspaceService workspaceService) {
+    this.azureConfig = azureConfig;
+    this.azureCloudContext = azureCloudContext;
+    this.crlService = crlService;
+    this.resource = resource;
+    this.workspaceService = workspaceService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    FlightMap inputMap = context.getInputParameters();
+    ApiAzureIpCreationParameters creationParameters =
+        inputMap.get(CREATION_PARAMETERS, ApiAzureIpCreationParameters.class);
+
+    ComputeManagerCow computeCow = crlService.getComputeManagerCow();
+
+    // Don't try to create it if it already exists. At this point the assumption is
+    // this is a redo and this step created it already.
+    PublicIpAddress existingIP =
+        computeCow
+            .computeManager()
+            .networkManager()
+            .publicIpAddresses()
+            .getByResourceGroup(azureCloudContext.getAzureResourceGroupId(), resource.getIpName());
+
+    if (existingIP == null) {
+      computeCow
+          .computeManager()
+          .networkManager()
+          .publicIpAddresses()
+          .define(resource.getIpName())
+          .withRegion(azureConfig.getDefaultRegion())
+          .withExistingResourceGroup(azureCloudContext.getAzureResourceGroupId())
+          .withDynamicIP()
+          .withTag("crl.integration", "true")
+          .create(
+              Defaults.buildContext(
+                  new CreatePublicIpRequestData(
+                      azureCloudContext.getAzureResourceGroupId(),
+                      resource.getIpName(),
+                      azureConfig.getDefaultRegion())));
+    } else {
+      logger.info("Ip {} already exists. Continuing.", resource.getIpName());
+    }
+
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    ComputeManagerCow computeCow = crlService.getComputeManagerCow();
+    computeCow
```

Comment:
When the flight starts, there are two cases:
1. The IP address needs to be created in the workspace.
2. The IP address already exists in the workspace. This should be an error, but you treat it as a re-run of the step. And on an error in a later step, your undo would delete the already existing IP address.

I suggest that you have another step that you run ahead of CreateAzureIpStep. It would detect and error out if the IP address already existed. That way this step knows that if the IP exists, it is because this step created it. The undo can safely delete it.

There is a race condition if there are two concurrent creators of the identical thing. That is pretty much true of all of the current WSM flights. We have not implemented a solution for that problem. If it becomes a real-world problem, we can fix it by serializing in WSM.

Reply:
I implemented it so we intelligently don't delete it if it already existed in undo, but I did it without another step for now, with the concurrency issue of the working map that you note.